src/ecrn_agent.erl

```diff
@@ -262,57 +339,77 @@ until_next_time(State, {{monthly, DoM, Period}, _What}) ->
+    true ->
+        case next_time(Period, Now) of
+            T when T > EndTime ->
+                until_tomorrow(State, StartTime);
+            T ->
+                T - Now
+        end
     end.
 
-%% @doc Calculates the last time in a given period.
--spec last_time(erlcron:period()) -> erlcron:seconds().
-last_time(Period) ->
-    hd(lists:reverse(lists:sort(resolve_period(Period)))).
-
-
 %% @doc Calculates the first time in a given period.
 -spec first_time(erlcron:period()) -> erlcron:seconds().
 first_time(Period) ->
-    hd(lists:sort(resolve_period(Period))).
+    case resolve_period(Period) of
+        [] -> 86400;
+        L  -> lists:min([FromTime || {FromTime, _ToTime, _RepeatSec} <- L])
+    end.
+
+%% @doc Calculates the last time in a given period.
+-spec last_time(erlcron:period()) -> erlcron:seconds().
+last_time(Period) ->
+    case resolve_period(Period) of
+        [] -> 86400;
+        L  -> F = fun({FromTime, _ToTime, 0}) ->
+                          FromTime;
+                     ({FromTime, ToTime, RepeatSec}) ->
+                          FromTime + ((ToTime-FromTime) div RepeatSec) * RepeatSec
+                  end,
+              lists:max([F(R) || R <- L])
+    end.
 
 %% @doc Calculates the first time in the given period after the given time.
 -spec next_time(erlcron:period(), erlcron:seconds()) -> erlcron:seconds().
 next_time(Period, Time) ->
-    R = lists:sort(resolve_period(Period)),
-    lists:foldl(fun(X, A) ->
-                        case X of
-                            T when T >= Time, T < A ->
-                                T;
-                            _ ->
-                                A
-                        end
-                end, 24*3600, R).
+    case resolve_period(Period) of
+        [] -> 86400;
+        L  -> F = fun({FromTime, _ToTime, _RepeatSec}) when Time =< FromTime ->
+                        FromTime;
+                     ({_FromTime, _ToTime, 0}) ->
+                        86400;
+                     ({FromTime, _ToTime, RepeatSec}) ->
+                        A1 = ((Time-FromTime) div RepeatSec) * RepeatSec,
+                        A  = if A1 < Time -> A1 + RepeatSec; true -> A1 end,
+                        FromTime + A
+                  end,
+              case [F(R) || R = {_, ToTm, _} <- L, Time =< ToTm] of
+                  [] -> 86400;
+                  LL -> lists:min(LL)
+              end
+    end.
 
 %% @doc Returns a list of times given a periodic specification.
--spec resolve_period([erlcron:period()] | erlcron:period()) -> [erlcron:seconds()].
-resolve_period([]) ->
-    [];
-resolve_period([H | T]) ->
-    resolve_period(H) ++ resolve_period(T);
-resolve_period({every, Duration, {between, TimeA, TimeB}}) ->
-    Period = resolve_dur(Duration),
+-spec resolve_period([erlcron:period()] | erlcron:period()) ->
+        [{erlcron:seconds(), erlcron:seconds(), erlcron:seconds()}].
+resolve_period([])      -> [];
```

Comment:
Maintain the same indentation as the other clauses.

Reply:
Fixed in the latest commit